internal/server/server.go

```diff
@@ -1,2 +1,179 @@
-// TODO: document the package.
+// This package containts server implementation to receive events in format described in
+// thermomatic/internal/client/reading.go.
 package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/spin-org/thermomatic/internal/client"
+	"github.com/spin-org/thermomatic/internal/imei"
+)
+
+const (
+	loginTimeout = 1 * time.Second
+	readTimeout = 2 * time.Second
+)
+
+var errReadTimeout = errors.New("read timeout")
+
+type Server struct {
+	create chan net.Conn
+	exit   chan net.Conn
+	clients map[net.Conn]struct{}
+
+	stdout io.Writer
+}
+
+func New(stdout io.Writer) *Server {
+	return &Server{
+		create: make(chan net.Conn),
+		exit:   make(chan net.Conn),
+
+		clients: make(map[net.Conn]struct{}),
+		stdout: stdout,
+	}
+}
+
+func (s *Server) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	lc := net.ListenConfig{}
+	l, err := lc.Listen(ctx, "tcp4", ":1337")
+	if err != nil {
+		return err
+	}
+
+	go s.accept(l)
+
+outer:
+	for {
+		select {
+		case conn := <- s.create:
+			log("%v %v new connection established\n", conn, conn.RemoteAddr())
+			s.clients[conn] = struct{}{}
+			go s.handleClient(conn)
+		case conn := <-s.exit:
+			log("%v %v connection is closing\n", conn, conn.RemoteAddr())
+			delete(s.clients, conn)
+			_ = conn.Close()
+		case <- ctx.Done():
+			log("graceful shutdown\n")
+			_ = l.Close()
+			for c, _ := range s.clients {
+				_ = c.Close()
+			}
+			break outer
+		}
+	}
+	s.shutdown()
+	return nil
+}
+
+func (s *Server) shutdown() {
+	if len(s.clients) == 0 {
+		return
+	}
+
+	for {
+		select {
+			case conn := <- s.create:
+				_ = conn.Close()
+			case conn := <- s.exit:
+				delete(s.clients, conn)
+				if len(s.clients) == 0 {
+					return
+				}
+		}
+	}
+}
+
+func (s *Server) accept(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
```

Comment:
Are there any errors a TCP/IP listener may recover from? Any that it may not?

Reply:
Good question. could not find in man pages (at least on mac) any errors from which a listener can be recovered. Probably better way to panic here but in this case grateful shutdown should be re-implemented

Reply:
from go sources:
> // Treat ECONNRESET and ECONNABORTED as temporary errors when
> // they come from calling accept. See issue 6163

Reply:
You may want to take a look [here](https://golang.org/src/net/http/server.go?s=90314:90360#L2856).